src/Containers/Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -339,68 +343,186 @@ public async Task<string> DownloadBlobAsync(string repository, Descriptor descri
+            // we use this data to update our internal state before each new iteration.
+            void UpdateStateFromResponse(HttpResponseMessage response, bool ignoreRange)
             {
-                var headers = patchResponse.Headers.ToString();
-                var detail = await patchResponse.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
-                string errorMessage = Resource.FormatString(nameof(Strings.BlobUploadFailed), $"Chunked PATCH {patchUri}", patchResponse.StatusCode, headers + Environment.NewLine + detail);
-                throw new ApplicationException(errorMessage);
+                var amountSent = ignoreRange ? null : ParseRangeAmount(response);
+                patchUri = GetNextLocation(response).Uri;
+                if (amountSent is not null)
+                {
+                    chunkStart = amountSent.Value + 1;
+                    // what the server actually read/processed may not align with the bytesRead from the content originally,
+                    // so seek to that position to ensure no gaps
+                    contents.Seek(chunkStart, SeekOrigin.Begin);
+                }
+                else
+                {
+                    chunkStart += bytesRead;
+                }
             }
 
-           localUploadUri = GetNextLocation(patchResponse);
+            static async Task<Action> LogError(HttpResponseMessage response, Uri patchUri, CancellationToken cancellationToken)
+            {
+                var headers = response.Headers.ToString();
+                var detail = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+                string errorMessage = Resource.FormatString(nameof(Strings.BlobUploadFailed), $"Chunked PATCH {patchUri}", response.StatusCode, headers + Environment.NewLine + detail);
+                return () => throw new ApplicationException(errorMessage);
+            }
 
-            patchUri = localUploadUri.Uri;
+            // when an error occurs, we can check on the state of the upload from the registry. notably this gives us the same
+            // location and range information as the success, so we can use this for our iteration processing.
+            async Task<Action> CheckReadAmountAndRetry(Uri patchUri, CancellationToken cancellationToken)
+            {
+                retryCount += 1;
+                var getResponse = await client.GetAsync(patchUri, cancellationToken).ConfigureAwait(false);
+                if (!getResponse.IsSuccessStatusCode || getResponse.StatusCode != HttpStatusCode.NoContent)
+                {
+                    // reset back to previous chunk so we can retry
+                    contents.Seek(chunkStart, SeekOrigin.Begin);
+                }
+                else
+                {
+                    UpdateStateFromResponse(getResponse, false);
+                }
+                return () => { };
+            }
```

Comment:
This is new, and also helps with resiliency - when we retry we might be in a state where the response from the server doesn't have position information (Range headers, etc). In that case, we can issue a GET to the upload URI to get a 'status' response that _does_ have Range headers, and use that to ensure we're in sync. This is why `UpdateStateFromResponse` is a separate method - so we can ensure the metadata is always acted upon the same way.